src/internal/client.ts

```diff
@@ -0,0 +1,383 @@
+    // Validate input params.
+    if (!isValidEndpoint(params.endPoint)) {
+      throw new errors.InvalidEndpointError(`Invalid endPoint : ${params.endPoint}`)
+    }
+    if (!isValidPort(params.port)) {
+      throw new errors.InvalidArgumentError(`Invalid port : ${params.port}`)
+    }
+    if (!isBoolean(params.useSSL)) {
+      throw new errors.InvalidArgumentError(
+        `Invalid useSSL flag type : ${params.useSSL}, expected to be of type "boolean"`,
+      )
+    }
+
+    // Validate region only if its set.
+    if (params.region) {
+      if (!isString(params.region)) {
+        throw new errors.InvalidArgumentError(`Invalid region : ${params.region}`)
+      }
+    }
+
+    const host = params.endPoint.toLowerCase()
+    let port = params.port
+    let protocol: string
+    let transport
+    let transportAgent: http.Agent
+    // Validate if configuration is not using SSL
+    // for constructing relevant endpoints.
+    if (params.useSSL) {
+      // Defaults to secure.
+      transport = https
+      protocol = 'https:'
+      port = port || 443
+      transportAgent = https.globalAgent
+    } else {
+      transport = http
+      protocol = 'http:'
+      port = port || 80
+      transportAgent = http.globalAgent
+    }
+
+    // if custom transport is set, use it.
+    if (params.transport) {
+      if (!isObject(params.transport)) {
+        throw new errors.InvalidArgumentError(
+          `Invalid transport type : ${params.transport}, expected to be type "object"`,
+        )
+      }
+      transport = params.transport
+    }
+
+    // if custom transport agent is set, use it.
+    if (params.transportAgent) {
+      if (!isObject(params.transportAgent)) {
+        throw new errors.InvalidArgumentError(
+          `Invalid transportAgent type: ${params.transportAgent}, expected to be type "object"`,
+        )
+      }
+
+      transportAgent = params.transportAgent
+    }
+
+    // User Agent should always following the below style.
+    // Please open an issue to discuss any new changes here.
+    //
+    //       MinIO (OS; ARCH) LIB/VER APP/VER
+    //
+    const libraryComments = `(${process.platform}; ${process.arch})`
+    const libraryAgent = `MinIO ${libraryComments} minio-js/${Package.version}`
+    // User agent block ends.
+
+    this.transport = transport
+    this.transportAgent = transportAgent
+    this.host = host
+    this.port = port
+    this.protocol = protocol
+    this.userAgent = `${libraryAgent}`
+
+    // Default path style is true
+    if (params.pathStyle === undefined) {
+      this.pathStyle = true
+    } else {
+      this.pathStyle = params.pathStyle
+    }
+
+    this.accessKey = params.accessKey ?? ''
+    this.secretKey = params.secretKey ?? ''
+    this.sessionToken = params.sessionToken
+    this.anonymous = !this.accessKey || !this.secretKey
+
+    if (params.credentialsProvider) {
+      this.credentialsProvider = params.credentialsProvider
+      void this.checkAndRefreshCreds()
```

Comment:
this may cause `unhandledrejection` ...

Reply:
since the (assume role) client initialization would fail, it should be fine. Later we can refactor credential providers to be like in other sdks.

Reply:
The consumer can always handle the error. @trim21 
Test script:
```javascript

const Minio = require('minio')
const { AssumeRoleProvider } = require('../dist/main/AssumeRoleProvider')

// Instantiate the minio client with the endpoint
// and access keys as shown below.
//Play instance

const hostName = 'localhost'
const apiPort = 22000
const stsEndPoint = `http://127.0.0.1:${apiPort}`
//Local testing

const bucketName = 'test-bucket'
const objectName = 'test-object'
const prefix = 'test'

const testDefaultClient = async () => {
  let minioClient = new Minio.Client({
    endPoint: hostName,
    port: apiPort,
    useSSL: false,
    accessKey: 'minio',
    secretKey: 'minio123',
  })
  const withoutRole = await minioClient.statObject(bucketName, objectName)
  console.log('Without Assume role:', withoutRole)
}
const testDynamicAssumeRoleClient = async () => {
  let minioClient = new Minio.Client({
    endPoint: hostName,
    port: apiPort,
    useSSL: false,
    accessKey: 'ro-user',
    secretKey: 'minio123',
  })

  try {
    let ap = new AssumeRoleProvider({
      stsEndpoint: stsEndPoint,
      accessKey: 'minio',
      secretKey: 'minio12345', //invalid credentials for testing error handling
    })

    await minioClient.setCredentialsProvider(ap)
    const withRole = await minioClient.statObject(bucketName, objectName)
    console.log('With Assume role ', withRole)
  } catch (err) {
    console.log('Error initializing assume role client..')
  }
}

const testInitializerAssumeRoleClient = async () => {
  //minioClient.traceOn()
  try {
    let asRoleProvider = new AssumeRoleProvider({
      stsEndpoint: stsEndPoint,
      /*accessKey: 'minio',
      secretKey: 'minio123',*/

      accessKey: 'ro-user',
      secretKey: 'minio123',
    })

    const asRoleClient = new Minio.Client({
      endPoint: hostName,
      port: apiPort,
      useSSL: false,
      accessKey: 'minio',
      secretKey: 'minio1234',//invalid creds for testing error hadling.
      credentialsProvider: asRoleProvider,
    })

    const withRole = await asRoleClient.statObject(bucketName, objectName)
    console.log(withRole)
  } catch (err) {
    console.log('Constructor Initialization of Assume Role::', new Date())
  }
}

//testDefaultClient()
//testDynamicAssumeRoleClient()
// testInitializerAssumeRoleClient()


```

Reply:
Iguess the last test will fail…

Reply:
Oh. yes i get what you are saying. it would continue with the s3 call and throw the s3 api error.

Reply:
Yes, I mean no error from `asRoleProvider` will be throwed.

Reply:
I think we can remove this line, and leave the problme to fix when refactoring the `requestSteam` method?

Reply:
Remove this line won't change client behaviour

Reply:
Yeah. how shall we go about this async initialization ?

Reply:
Sure. we can remove it. yes. it makes sense. Thank you for the suggestion . :+1: